lib/client.c

```diff
@@ -1071,23 +1074,111 @@ SF_STMT *STDCALL snowflake_stmt(SF_CONNECT *sf) {
     return sfstmt;
 }
 
+/* Leave the memory management of the bind input to the user
+ * The user will allocate memory for the SF_BIND_INPUT and
+ * pass it to the init function for sane initialization
+ */
+void STDCALL snowflake_bind_input_init(SF_BIND_INPUT * input)
+{
+    if (!input)
+    {
+#ifdef SF_DEBUG
+        printf("[SF_ERROR] snowflake_bind_input: bad param. user is supposed to allocate memory for input\n");
+#endif
+        return;
+    }
+    input->idx = 0;
+    input->name = NULL;
+    input->value = NULL;
+
+    return;
+}
+
 void STDCALL snowflake_stmt_term(SF_STMT *sfstmt) {
     if (sfstmt) {
         _snowflake_stmt_reset(sfstmt);
         SF_FREE(sfstmt);
     }
 }
 
+PARAM_TYPE STDCALL _sf_get_param_style(const SF_BIND_INPUT *input)
+{
+    PARAM_TYPE retval = INVALID_PARAM_TYPE;
+
+    if (!input)
+    {
+        goto done;
+    }
+    if (input->name && !input->idx)
+    {
+        retval = NAMED;
+    }
+    else
+    {
+        retval = POSITIONAL;
+    }
+
+    done:
+    return retval;
+}
+
+PARAM_TYPE STDCALL _sf_get_current_param_style(const SF_STMT *sfstmt)
+{
+    PARAM_STORE *ps = NULL;
+    if (!sfstmt || !sfstmt->params)
+    {
+        return INVALID_PARAM_TYPE;
+    }
+
+    ps = (PARAM_STORE *)sfstmt->params;
+
+    return ps->param_style;
+}
+
 SF_STATUS STDCALL snowflake_bind_param(
     SF_STMT *sfstmt, SF_BIND_INPUT *sfbind) {
+
+    SF_RET_CODE retcode;
     if (!sfstmt) {
         return SF_STATUS_ERROR_STATEMENT_NOT_EXIST;
     }
     clear_snowflake_error(&sfstmt->error);
     if (sfstmt->params == NULL) {
-        sfstmt->params = sf_array_list_init();
+        sf_param_store_init(_sf_get_param_style(sfbind),
+                            &sfstmt->params);
+    }
+    retcode = sf_param_store_set(sfstmt->params, sfbind, sfbind->idx, sfbind->name);
+    if (retcode == SF_RET_CODE_DUPLICATES)
+    {
+        return SF_STATUS_SUCCESS;
+    }
+    else if (retcode == SF_RET_CODE_ERROR)
+    {
+        return SF_STATUS_ERROR_OTHER;
+    }
+    /*
+     * For named parameters, populate the name list in the
+     * sfstmt to help extract the params from the treemap
+     * during execute
+     */
+    if (_sf_get_current_param_style(sfstmt) == NAMED)
+    {
+        sfstmt->name_list = SF_REALLOC(sfstmt->name_list,
```

Comment:
Could we grow this list more than 1 element at a time? It seems like the realloc function will get called n times (where n is the number of input parameters for a sql query).